src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,151 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        linkLast(value);
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("error " + index);
+        }
+        if (index == size) {
+            linkLast(value);
+        } else if (isEmpty()) {
+            add(value);
+
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head = newNode;
+            newNode.next.prev = newNode;
+        } else {
+            linkBefore(value, getNodeByIndex(index));
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        list.forEach(this::add);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexInBounds(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexInBounds(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexInBounds(index);
+        Node<T> node = getNodeByIndex(index);
+        unlink(node);
+        return node.value;
     }
 
     @Override
     public boolean remove(T object) {
+        T value;
+        for (int i = 0; i < size; i++) {
+            value = getNodeByIndex(i).value;
+            if (value == object || value != null && value.equals(object)) {
+                remove(i);
```

Comment:
calling remove from here will cause searching for that node again via `getNodeByIndex()` - redundant complexity

since you've already found the desired node - why not unlinking it directly from here?